Updated M26 table3 setup to better support larger formatted values (dictionary + overflow)

WiiExpertise opened the pull request:
- Updated zstd table3 strategy to utilize dictionary support added to Node's zlib module.
- Resolves situations where larger formatted values did not fit within the maxLength of the table3 field when compressed without the dictionary
- Added a corresponding test that tests setting a larger formatted value
- Added a fallback to attempt compressing with the max zstd compression level if the regular compression level is still enough. Only used on an as needed basis for performance reasons.
- Removed `@toondepauw/node-zstd` dependency and replaced with native zlib functions since it now supports the use of dictionaries
- **NOTE:** In order for this to work, Node version must be >= 22.19.0 or >= 24.6.0, as older versions do not contain dictionary support for zstd. Updated engine requirement accordingly.

- Added support for reading M26 table3 fields that utilize an overflow record reference as well as writing table3 formatted values that are too large to fit within the maxLength of the field.
- Added corresponding tests for reading/writing fields with overflow data
- **Limitation:** Cannot handle cases where the overflow reference points to a record in a different table. These situations should be exceedingly rare, however, as the table would have to run out of space for it to occur naturally.

WiiExpertise committed:
Updated M26 visuals strategy to utilize zstd dictionary for writing

WiiExpertise committed:
Add support for handling overflow table3 records in M26

bep713 commented on src/FranchiseFileTable.js:
```diff
@@ -484,8 +484,20 @@ class FranchiseFileTable extends EventEmitter {
+            // Check if this record has an overflow record reference
+            const overflowRef = record['Overflow'];
+            const overflowTableId = utilService.bin2dec(overflowRef.substring(0, 15));
+            const overflowTableRow = utilService.bin2dec(overflowRef.substring(15));
```
```suggestion
            const { tableId: overflowTableId, rowNumber: overflowTableRow } = utilService.getReferenceData(overflowRef);
```

bep713 commented on src/FranchiseFileField.js:
```diff
@@ -185,6 +185,12 @@ class FranchiseFileField {
     }
     ;
+
+    /** @returns {FranchiseFIleRecord} */
```
```suggestion
    /** @returns {FranchiseFileRecord} */
```

bep713 commented on src/FranchiseFileTable.js:
```diff
@@ -484,8 +484,20 @@ class FranchiseFileTable extends EventEmitter {
+            if(overflowTableId !== 0 && overflowTableId === header.tableId)
+            {
+                overflowRecord = records[overflowTableRow];
+            }
```
Extract this or part of it to a util function - can reuse in FranchiseFileTable3Field as well

WiiExpertise replied:
Resolved

bep713 commented on src/FranchiseFileTable3Field.js:
```diff
@@ -86,6 +103,75 @@ class FranchiseFileTable3Field {
+            if(nextRecordIndex != table.header.recordCapacity)
+            {
+                // Update the Overflow reference of the current record to point to the new overflow record
+                const reference = utilService.dec2bin(table.header.tableId, 15) + utilService.dec2bin(nextRecordIndex, 17);
```
```suggestion
                const reference = utilService.getBinaryReferenceData(table.header.tableId, nextRecordIndex);
```

WiiExpertise replied:
Resolved

bep713 commented on src/FranchiseFileTable3Field.js:
```diff
@@ -86,6 +103,75 @@ class FranchiseFileTable3Field {
+
+        // If the referenced record is in the current table, clear it out and empty it
+        const overflowTableId = utilService.bin2dec(overflowRef.substring(0, 15));
+        const overflowTableRow = utilService.bin2dec(overflowRef.substring(15));
```
use `utilService.getReferenceData` here instead

WiiExpertise replied:
Should be resolved in latest commit.

bep713 reviewed (commented):
👍 just have a few comments. Good catch with the overflow data.

WiiExpertise committed:
Cleaned up code

WiiExpertise commented:
Implemented suggested changes in latest commit. Moved the overflow record getting from the reference to a utilService function that takes in the record object and returns either null or the overflow record object.
